### PR TITLE
ISPN-12657 Cross-Site automatic state transfer

### PR DESCRIPTION
**Pull request by pruivo**
https://issues.redhat.com/browse/ISPN-12657

### CODE REVIEW

File: cli/src/main/java/org/infinispan/cli/commands/rest/Site.java [hunk 1]
```diff
@@ -274,4 +282,42 @@ protected CommandResult exec(ContextAwareCommandInvocation invocation) throws Co
...
+      protected boolean help;
+
+      @Option(shortName = 'c', completer = CacheCompleter.class, description = "The cache name.")
+      String cache;
+
+      @Option(shortName = 's', required = true, completer = SiteCompleter.class, description = "The remote backup name.")
+      String site;
+
+      @Option(shortName = 'm', completer = XSiteStateTransferModeCompleter.class, description = "The state transfer mode to set.")
+      protected String mode;
```
Comment (ryanemerson) [inline cli/src/main/java/org/infinispan/cli/commands/rest/Site.java:299]: Should this be `required`?

File: cli/src/main/resources/help/site.adoc [hunk 2]
```diff
@@ -67,6 +74,12 @@ Cancels the operation to receive state from `NYC`.
 `site clear-push-state-status --cache=myCache` +
 Clears the status of the push state operation for `mycache`.
 
+`site state-transfer-mode --cache=myCache --site=NYC` +
+Retrieves the state transfer mode for the cache `mycache` and remote backup `NYC`
+
+`site state-transfer-mode --cache=myCache --site=NYC --mode=AUTO` +
+Sets the state transfer mode to `AUTO` for the cache `mycache` and remote backup `NYC`
```
Comment (ryanemerson) [inline cli/src/main/resources/help/site.adoc:81]: It seems strange to configure the value by just setting a `--mode=AUTO` flag. Shouldn't we have dedicated commands for get & set? e.g.

```
site state-transfer-mode get --cache=myCache --site=NYC
site state-transfer-mode set --cache=myCache --site=NYC AUTO
```

File: core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java [hunk 3]
```diff
@@ -1229,6 +1226,16 @@ private void updateSitesView(Collection<String> sitesUp, Collection<String> site
...
          viewUpdateLock.unlock();
       }
+      if (sitesUp.isEmpty()) {
+         return;
+      }
+      if (isCoordinator()) {
+         globalXSiteAdminOperations.onSitesUp(sitesUp);
+      } else {
+         //for the case where the coordinator isn't the site master.
+         //TODO improve this by checking if the coordinator is a site master or not.
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java:1236]: Is this meant to be fixed in this PR? If not, let's create a Jira and reference the id here.

File: core/src/main/java/org/infinispan/xsite/XSiteAdminOperations.java [hunk 4]
```diff
@@ -318,6 +325,48 @@ public final String getSendingSiteName() {
       return stateTransferManager.getSendingSiteName();
    }
 
+   public final CompletionStage<String> asyncGetStateTransferMode(String site) {
+      authorizer.checkPermission(AuthorizationPermission.ADMIN);
+      //check local first
+      if (stateTransferManager.stateTransferMode(site) == XSiteStateTransferMode.MANUAL) {
+         return CompletableFuture.completedFuture("MANUAL");
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/xsite/XSiteAdminOperations.java:332]: `XSiteStateTransferMode.MANUAL.toString()`?

File: server/rest/src/main/java/org/infinispan/rest/resources/XSiteResource.java [hunk 5]
```diff
@@ -113,6 +117,12 @@ public Invocations getInvocations() {
             .invocation().methods(POST).path("/v2/caches/{cacheName}/x-site/backups/{site}").withAction("cancel-receive-state")
                .permission(AuthorizationPermission.ADMIN).name("XSITE CANCEL RECEIVE STATE").auditContext(AuditContext.CACHE)
                .handleWith(this::cancelReceiveState)
+            .invocation().methods(GET).path("/v2/caches/{cacheName}/x-site/backups/{site}/state-transfer-mode")
+               .permission(AuthorizationPermission.ADMIN).name("XISTE GET STATE TRANSFER MODE").auditContext(AuditContext.CACHE)
```
Comment (ryanemerson) [inline server/rest/src/main/java/org/infinispan/rest/resources/XSiteResource.java:121]: `/XISTE/XSITE` :slightly_smiling_face:

File: cli/src/main/resources/help/site.adoc [hunk 6]
```diff
@@ -37,6 +39,11 @@ Specifies a cache.
 *--site*='SITE_NAME'::
 Specifies a backup location.
 
+STATE TRANSFER MODE OPTIONS
+---------------------------
+
+*--mode*='MODE'::
+Sets the state transfer mode (`MANUAL` or `AUTO`)
```
Comment (oraNod) [inline cli/src/main/resources/help/site.adoc:46]:
```suggestion
Sets the state transfer mode. Values are `MANUAL` or `AUTO`.
```

Comment (oraNod) [reply cli/src/main/resources/help/site.adoc:46]: @pruivo Is `MANUAL` the default?

File: server/rest/src/main/java/org/infinispan/rest/resources/XSiteResource.java [hunk 7]
```diff
@@ -246,50 +258,93 @@ public Invocations getInvocations() {
...
+      String mode = request.getParameter("mode");
+      if (mode == null) {
+         return completedFuture(responseBuilder.status(BAD_REQUEST).build());
+      }
+      //check if site exists
+      final String site = request.variables().get("site");
+      Optional<XSiteAdminOperations> xsiteAdminOpt = getXSiteAdminAndCheckSite(request, responseBuilder);
+      return xsiteAdminOpt.<CompletionStage<RestResponse>>map(ops ->
+            ops.asyncSetStateTransferMode(site, mode)
+                  .thenApply(ok -> responseBuilder.status(ok ? OK : NOT_MODIFIED).build()))
```
Comment (ryanemerson) [inline server/rest/src/main/java/org/infinispan/rest/resources/XSiteResource.java:318]: Is `NOT_MODIFIED` sufficient? Or should we include a message explaining that the requested mode is not supported on the remote site?

File: cli/src/main/resources/help/site.adoc [hunk 8]
```diff
@@ -67,6 +74,12 @@ Cancels the operation to receive state from `NYC`.
 `site clear-push-state-status --cache=myCache` +
 Clears the status of the push state operation for `mycache`.
 
+`site state-transfer-mode --cache=myCache --site=NYC` +
+Retrieves the state transfer mode for the cache `mycache` and remote backup `NYC`
```
Comment (oraNod) [inline cli/src/main/resources/help/site.adoc:78]:
```suggestion
Retrieves the state transfer mode for `mycache` to `NYC`.
```

File: commons/all/src/main/java/org/infinispan/configuration/cache/XSiteStateTransferMode.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+/**
+ * Cross site state transfer mode.
+ *
+ * @author Pedro Ruivo
+ * @since 12.1
+ */
+public enum XSiteStateTransferMode {
+   /**
+    * The cross-site state transfer is triggered manually (via CLI/JMX/REST).
```
Comment (oraNod) [inline commons/all/src/main/java/org/infinispan/configuration/cache/XSiteStateTransferMode.java:11]:
```suggestion
    * Cross-site state transfer is triggered manually via CLI, JMX, or REST.
```

File: commons/all/src/main/java/org/infinispan/configuration/cache/XSiteStateTransferMode.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    */
+   MANUAL,
+   /**
+    * The cross-site state transfer is triggered automatically.
```
Comment (oraNod) [inline commons/all/src/main/java/org/infinispan/configuration/cache/XSiteStateTransferMode.java:15]:
```suggestion
    * Cross-site state transfer is triggered automatically.
```

File: core/src/main/java/org/infinispan/configuration/cache/XSiteStateTransferConfigurationBuilder.java [hunk 11]
```diff
@@ -78,22 +86,34 @@ public final XSiteStateTransferConfigurationBuilder timeout(long timeout) {
    }
 
    /**
-    * The maximum number of retries when a push state command fails. A value &le; 0 (zero) mean that the command will not
-    * retry. Default value is 30.
+    * The maximum number of retries when a push state command fails. A value &le; 0 (zero) mean that the command will
```
Comment (oraNod) [inline core/src/main/java/org/infinispan/configuration/cache/XSiteStateTransferConfigurationBuilder.java:89]:
```suggestion
    * The maximum number of retries when a push state command fails. A value &le; 0 (zero) means that the command does
```

File: core/src/main/java/org/infinispan/configuration/cache/XSiteStateTransferConfigurationBuilder.java [hunk 12]
```diff
@@ -78,22 +86,34 @@ public final XSiteStateTransferConfigurationBuilder timeout(long timeout) {
...
+    * not retry. Default value is 30.
     */
    public final XSiteStateTransferConfigurationBuilder maxRetries(int maxRetries) {
       attributes.attribute(MAX_RETRIES).set(maxRetries);
       return this;
    }
 
    /**
-    * The waiting time (in milliseconds) between each retry. The value should be &gt; 0 (zero). Default value is 2 seconds.
+    * The waiting time (in milliseconds) between each retry. The value should be &gt; 0 (zero). Default value is 2
```
Comment (oraNod) [inline core/src/main/java/org/infinispan/configuration/cache/XSiteStateTransferConfigurationBuilder.java:98]:
```suggestion
    * The wait time, in milliseconds, between each retry. The value should be &gt; 0 (zero). Default value is 2
```

File: core/src/test/java/org/infinispan/xsite/statetransfer/XSiteAutoStateTransferTest.java [hunk 13]
```diff
@@ -0,0 +1,448 @@
...
+         assertNull(cache(1, 0).get(TestingUtil.k(method, i)));
+      }
+
+      SiteMasterController controller = findSiteMaster();
+
+      //let the state command go through
+      controller.getRpcManager().excludeCommands(XSiteStatePushCommand.class);
+      CompletableFuture<ControlledRpcManager.BlockedRequest<XSiteAutoTransferStatusCommand>> req1 = controller
+            .getRpcManager().expectCommandAsync(XSiteAutoTransferStatusCommand.class);
+      CompletableFuture<ControlledRpcManager.BlockedRequest<XSiteBringOnlineCommand>> req2 = controller.getRpcManager()
```
Comment (ryanemerson) [inline core/src/test/java/org/infinispan/xsite/statetransfer/XSiteAutoStateTransferTest.java:200]: Can you split this after `=` to make it a bit more readable.

File: core/src/main/java/org/infinispan/configuration/cache/XSiteStateTransferConfigurationBuilder.java [hunk 14]
```diff
@@ -78,22 +86,34 @@ public final XSiteStateTransferConfigurationBuilder timeout(long timeout) {
...
     */
    public final XSiteStateTransferConfigurationBuilder waitTime(long waitingTimeBetweenRetries) {
       attributes.attribute(WAIT_TIME).set(waitingTimeBetweenRetries);
       return this;
    }
 
+   /**
+    * The cross-site state transfer mode.
+    * <p>
+    * If set to {@link XSiteStateTransferMode#AUTO}, Infinispan will automatic start the state transfer when it detects
```
Comment (oraNod) [inline core/src/main/java/org/infinispan/configuration/cache/XSiteStateTransferConfigurationBuilder.java:109]:
```suggestion
    * If set to {@link XSiteStateTransferMode#AUTO}, Infinispan automatically starts state transfer when it detects
```

File: core/src/main/java/org/infinispan/configuration/cache/XSiteStateTransferConfigurationBuilder.java [hunk 15]
```diff
@@ -78,22 +86,34 @@ public final XSiteStateTransferConfigurationBuilder timeout(long timeout) {
...
+    * a new view with the backup site and it was previously offline.
```
Comment (oraNod) [inline core/src/main/java/org/infinispan/configuration/cache/XSiteStateTransferConfigurationBuilder.java:110]:
```suggestion
    * a new view for a backup location that was previously offline.
```

File: core/src/main/java/org/infinispan/util/logging/Log.java [hunk 16]
```diff
@@ -2151,4 +2151,14 @@ CacheConfigurationException storeConfiguredHasBothReadAndWriteOnly(String storeC
...
    void topologyChangedPartitionHandlingTxCompletion(GlobalTransaction globalTransaction);
+
+   @Message(value = "Cross-Site state transfer mode cannot be null.", id = 633)
+   CacheConfigurationException invalidXSiteStateTransferMode();
+
+   @Message(value = "Cross-Site automatic state transfer is not compatible with SYNC backup strategy.", id = 634)
+   CacheConfigurationException autoXSiteStateTransferModeNotAvailableInSync();
+
+   @LogMessage(level = WARN)
+   @Message(value = "[%s] Failed to receive a response from all node. Automatic cross-site state transfer to site '%s' is not started.", id = 635)
```
Comment (oraNod) [inline core/src/main/java/org/infinispan/util/logging/Log.java:2162]:
```suggestion
   @Message(value = "[%s] Failed to receive a response from any nodes. Automatic cross-site state transfer to site '%s' is not started.", id = 635)
```

File: core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateTransferManager.java [hunk 17]
```diff
@@ -112,4 +114,29 @@
     * @return The {@link XSiteStateConsumer} instance.
     */
    XSiteStateConsumer getStateConsumer();
+
+   /**
+    * Starts (if required) the cross-site state transfer for the remote sites.
```
Comment (oraNod) [inline core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateTransferManager.java:119]:
```suggestion
    * Starts cross-site state transfer for the remote sites, if required.
```

File: core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateTransferManager.java [hunk 18]
```diff
@@ -112,4 +114,29 @@
...
+   /**
+    * @param site The remote site.
+    * @return The {@link XSiteStateTransferMode} configured for the remote site.
+    */
+   XSiteStateTransferMode stateTransferMode(String site);
+
+   /**
+    * Sets the {@link XSiteStateTransferMode} to the remote site.
+    * <p>
+    * If the configure for the remote site does not support the {@link XSiteStateTransferMode}, then this method returns
```
Comment (oraNod) [inline core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateTransferManager.java:134]:
```suggestion
    * If the configuration for the remote site does not support the {@link XSiteStateTransferMode}, then this method returns
```

File: core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateTransferManagerImpl.java [hunk 19]
```diff
@@ -229,6 +215,72 @@ public XSiteStateConsumer getStateConsumer() {
...
+            bringSiteOnline(site).thenRun(() -> asyncStartPushState(status));
+         } else {
+            Log.XSITE.debugf("[%s] Cross-Site state transfer not required for site '%s'", cacheName, site);
+         }
+      });
+   }
+
+   private boolean skipAutomaticStateTransferEnabled(String site, RemoteSiteStatus status) {
+      if (status == null) {
+         Log.XSITE.debugf("[%s] Cross-Site automatic state transfer not started for site '%s'. It is not a backup of this cache", cacheName, site);
```
Comment (oraNod) [inline core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateTransferManagerImpl.java:255]:
```suggestion
         Log.XSITE.debugf("[%s] Cross-Site automatic state transfer not started for site '%s'. It is not a backup location for this cache", cacheName, site);
```

File: cli/src/main/resources/help/site.adoc (hunk 6 above)
Comment (oraNod) [reply cli/src/main/resources/help/site.adoc:46]: Nevermind. I found this in the schema.

File: core/src/main/resources/schema/infinispan-config-12.1.xsd [hunk 20]
```diff
@@ -2158,6 +2158,14 @@
               </xs:documentation>
             </xs:annotation>
           </xs:attribute>
+          <xs:attribute name="mode" type="tns:xsite-state-transfer-mode" default="${XSiteStateTransfer.mode}">
+            <xs:annotation>
+              <xs:documentation>
+                When set to "auto", Infinispan will automatic start the state transfer when it detects the backup site
```
Comment (oraNod) [inline core/src/main/resources/schema/infinispan-config-12.1.xsd:2164]:
```suggestion
                Controls whether cross-site state transfer happens manually on user action or automatically when Infinispan
```

File: core/src/main/resources/schema/infinispan-config-12.1.xsd [hunk 21]
```diff
@@ -2158,6 +2158,14 @@
...
+                is back online. Default is "manual".
```
Comment (oraNod) [inline core/src/main/resources/schema/infinispan-config-12.1.xsd:2165]:
```suggestion
                detects that backup locations come online. The default is "manual".
```

File: core/src/main/resources/schema/infinispan-config-12.1.xsd [hunk 22]
```diff
@@ -2895,4 +2903,25 @@
       </xs:enumeration>
     </xs:restriction>
   </xs:simpleType>
+  <xs:simpleType name="xsite-state-transfer-mode">
+    <xs:restriction base="xs:token">
+      <xs:enumeration value="MANUAL">
+        <xs:annotation>
+          <xs:documentation>
+            Manual state transfer mode. If a backup site is offline, it needs manual intervention to bring it back
```
Comment (oraNod) [inline core/src/main/resources/schema/infinispan-config-12.1.xsd:2911]:
```suggestion
            Users must manually bring backup locations online and start state transfer operations.
```

File: core/src/main/resources/schema/infinispan-config-12.1.xsd [hunk 23]
```diff
@@ -2895,4 +2903,25 @@
...
+            online and to trigger the state transfer.
```
Comment (oraNod) [inline core/src/main/resources/schema/infinispan-config-12.1.xsd:2912]:
```suggestion

```

File: core/src/main/resources/schema/infinispan-config-12.1.xsd [hunk 24]
```diff
@@ -2895,4 +2903,25 @@
...
+          </xs:documentation>
+        </xs:annotation>
+      </xs:enumeration>
+      <xs:enumeration value="AUTO">
+        <xs:annotation>
+          <xs:documentation>
+            Automatic state transfer mode. If a backup site is offline, Infinispan brings it back online and triggers
```
Comment (oraNod) [inline core/src/main/resources/schema/infinispan-config-12.1.xsd:2919]:
```suggestion
            Infinispan automatically brings backup locations online and starts state transfer operations when the remote site
```

File: core/src/main/resources/schema/infinispan-config-12.1.xsd [hunk 25]
```diff
@@ -2895,4 +2903,25 @@
...
+            the state transfer when the backup site connection is stabilished.
```
Comment (oraNod) [inline core/src/main/resources/schema/infinispan-config-12.1.xsd:2920]:
```suggestion
            connection is stable.
```

File: documentation/src/main/asciidoc/topics/proc_cli_xsite_state_transfer_mode.adoc [hunk 26]
```diff
@@ -0,0 +1,19 @@
+[id='cli_xsite_push-{context}']
+= Set the Cross Site state transfer mode
```
Comment (oraNod) [inline documentation/src/main/asciidoc/topics/proc_cli_xsite_state_transfer_mode.adoc:2]:
```suggestion
= Configuring Cross-Site State Transfer Modes
```

File: documentation/src/main/asciidoc/topics/proc_cli_xsite_state_transfer_mode.adoc [hunk 27]
```diff
@@ -0,0 +1,19 @@
...
+Enable or disable the automatic cross site state transfer.
```
Comment (oraNod) [inline documentation/src/main/asciidoc/topics/proc_cli_xsite_state_transfer_mode.adoc:3]:
```suggestion
You can configure cross-site state transfer operations to happen automatically when {brandname} detects that backup locations come online or manually, which is the default.
```

File: documentation/src/main/asciidoc/topics/proc_cli_xsite_state_transfer_mode.adoc [hunk 28]
```diff
@@ -0,0 +1,19 @@
...
+
+.Procedure
+
+. Create a CLI connection to {brandname}.
+. Use the [command]`site` command to retrieve or set the state transfer mode, as in the following example:
```
Comment (oraNod) [inline documentation/src/main/asciidoc/topics/proc_cli_xsite_state_transfer_mode.adoc:8]:
```suggestion
. Use the [command]`site` command to configure state transfer modes, as in the following example:
```

File: documentation/src/main/asciidoc/topics/proc_cli_xsite_state_transfer_mode.adoc [hunk 29]
```diff
@@ -0,0 +1,19 @@
...
++
+----
+# retrives the current state transfer mode
```
Comment (oraNod) [inline documentation/src/main/asciidoc/topics/proc_cli_xsite_state_transfer_mode.adoc:11]:
```suggestion
# Retrieve the current state transfer mode
```

File: documentation/src/main/asciidoc/topics/proc_cli_xsite_state_transfer_mode.adoc [hunk 30]
```diff
@@ -0,0 +1,19 @@
...
+//containers/default]> site state-transfer-mode --cache=cacheName --site=NYC
+"MANUAL"
+
+# sets the state transfer mode
```
Comment (oraNod) [inline documentation/src/main/asciidoc/topics/proc_cli_xsite_state_transfer_mode.adoc:15]:
```suggestion
# Configure automatic state transfer operations for a cache and backup location.
```

File: documentation/src/main/asciidoc/topics/con_xsite_state_transfer.adoc [hunk 31]
```diff
@@ -48,3 +48,18 @@ Run `help clearcache` from the CLI for command details and examples.
 ====
 
 * link:{rest_docs}#rest_v2_clear_cache[Clearing Caches with the REST API]
+
+== Automatic State Transfer
+
+For asynchronous backup strategy, automatic state transfer is available.
```
Comment (oraNod) [inline documentation/src/main/asciidoc/topics/con_xsite_state_transfer.adoc:54]:
```suggestion
When using the asynchronous backup strategy, {brandname} can automatically perform cross-site state transfer operations.
```

File: documentation/src/main/asciidoc/topics/con_xsite_state_transfer.adoc [hunk 32]
```diff
@@ -48,3 +48,18 @@ Run `help clearcache` from the CLI for command details and examples.
...
+When a network connection is restored, {brandname} can initiate the bi-directional state transfer,
```
Comment (oraNod) [inline documentation/src/main/asciidoc/topics/con_xsite_state_transfer.adoc:55]:
```suggestion
When it detects that a backup location has come back online and the network connection is stable, {brandname} can initiate bi-directional state transfer operations.
```

File: documentation/src/main/asciidoc/topics/con_xsite_state_transfer.adoc [hunk 33]
```diff
@@ -48,3 +48,18 @@ Run `help clearcache` from the CLI for command details and examples.
...
+i.e. from **LON** to **NYC** and vice-versa, at the same time.
```
Comment (oraNod) [inline documentation/src/main/asciidoc/topics/con_xsite_state_transfer.adoc:56]:
```suggestion
For example, cross-site state transfer can automatically backup from **LON** to **NYC** and vice-versa at the same time.
```

File: documentation/src/main/asciidoc/topics/con_xsite_state_transfer.adoc [hunk 34]
```diff
@@ -48,3 +48,18 @@ Run `help clearcache` from the CLI for command details and examples.
...
+
+[NOTE]
+====
+This behaviour is disabled by default and must be configured.
```
Comment (oraNod) [inline documentation/src/main/asciidoc/topics/con_xsite_state_transfer.adoc:60]:
```suggestion
By default, automatic cross-site state transfer is disabled.
```

File: documentation/src/main/asciidoc/topics/proc_xsite_configuring_state_transfer.adoc [hunk 35]
```diff
@@ -0,0 +1,30 @@
+[id='configure_state_transfer-{context}']
+= Configuring Cross-Site State Transfer
+
+Configure the state transfer parameters and mode.
+
+.Procedure
+* Use `<state-transfer>` tag to configure the state transfer parameters and mode.
```
Comment (oraNod) [inline documentation/src/main/asciidoc/topics/proc_xsite_configuring_state_transfer.adoc:7]:
```suggestion
* Use `<state-transfer>` to configure state transfer operations.
```

File: documentation/src/main/asciidoc/topics/proc_xsite_configuring_state_transfer.adoc [hunk 36]
```diff
@@ -0,0 +1,30 @@
...
+
+.Example
```
Comment (oraNod) [inline documentation/src/main/asciidoc/topics/proc_xsite_configuring_state_transfer.adoc:9]:
```suggestion
For example, the following configuration automatically performs cross-site state transfer with custom settings:
```

File: documentation/src/main/asciidoc/topics/proc_xsite_configuring_state_transfer.adoc [hunk 37]
```diff
@@ -0,0 +1,30 @@
...
+* `chunk-size`: The number of cache entries to batch in each transfer.
+* `timeout`: The time, in milliseconds, to wait for the backup site acknowledge the state chunk received and applied.
+* `max-retires`: Sets the maximum number of retry attempts for push state failures.
+Specify a value of 0 (zero) to disable retry attempts.
+* `wait-time`: The time, in milliseconds, to wait between retry attempts for push state failures.
+The value must be greater than 0 (zero).
+* `mode`: Possible values are `MANUAL` and `AUTO`.
+`AUTO` is only available for `ASYNC` backup strategy and starts the state transfer automatically when a connection
+to a remote site is established.
+If `MANUAL`, the state transfer must be triggered manually.
```
Comment (oraNod) [inline documentation/src/main/asciidoc/topics/proc_xsite_configuring_state_transfer.adoc:24]: @pruivo Can we please delete the list of state transfer configuration properties here? This is reference info and is available in the schema.

Comment (pruivo) [reply documentation/src/main/asciidoc/topics/proc_xsite_configuring_state_transfer.adoc:24]: sure

File: server/rest/src/main/java/org/infinispan/rest/resources/XSiteResource.java (hunk 7 above)
Comment (pruivo) [reply server/rest/src/main/java/org/infinispan/rest/resources/XSiteResource.java:318]: Our REST client ignores the body if NOT_MODIFIED is returned and I'm hoping the users read the docs before invoking commands randomly.

File: core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java (hunk 3 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java:1236]: It was just an idea I had today. Not sure if JGroups exposes the info required or not.

File: cli/src/main/resources/help/site.adoc (hunk 2 above)
Comment (pruivo) [reply cli/src/main/resources/help/site.adoc]: ok

**Comment by pruivo** [comment]
rebased & updated

**Comment by oraNod** [comment]
@pruivo Hey, I noticed a few other edits with the docs here. I added a commit so please take a look before forcing a push. :wink:

**Review by ryanemerson — APPROVED** [review]

**Review by oraNod — APPROVED** [review]
LGTM

**Comment by ryanemerson** [comment]
Thanks @pruivo

**Pull request merged**